### PR TITLE
Hashmod - Allow transactions to have attributes from the CSV (WIP)

### PR DESCRIPTION
**Pull request by CrucisGamma**
### Description of the Change

Modifications to the experimental tool option Hashmod (adding in allowing Transactions to have an attribute from one of the vertices)
Modifications to the Time Line tool (initial only - looking for a bit of feedback on these a bit more)

### Alternate Designs

I would have liked to make the time tool have node based time information (may revisit this).  
My view of hashmod is that you should be able to ingest a CSV file and have it automatically construct a graph with transactions and vertices with the data that you would like to have in it with minimal effort.  

### Why Should This Be In Core?

Having more ways to present data and making things simpler and more configurable is desirable.

### Benefits

Better code in hashmod and more functionality for presentations in the time line tool

### Possible Drawbacks

UI interaction with the play buttons

### Verification Process

Still a work in progress

### Applicable Issues

#1363

### CODE REVIEW

**Comment by arcturus2** [comment]
@CrucisGamma I did a bit of testing with the timeline changes and I wasn't able to get it to work. It seemed to lock up actually. I'm happy to merge in your Hashmod part but can you put that in a separate PR?

File: CoreTimelineView/src/au/gov/asd/tac/constellation/views/timeline/TimelineTopComponent.java [hunk 1]
```diff
@@ -213,7 +208,70 @@ public void setExtents(final double lowerTimeExtent, final double upperTimeExten
...
+        originalLowerTime = getTimelineLowerTimeExtent();
+        originalUpperTime = getTimelineUpperTimeExtent();
+        videoParamsSet = true;
+
+        while (videoPos < 100 && videoParamsSet) {
+            double newLowerTime = (originalUpperTime - originalLowerTime) / 101 * videoPos + originalLowerTime;
+            double newUpperTime = (originalUpperTime - originalLowerTime) / 101 * (videoPos + 1) + originalLowerTime;
+            setExtents(newLowerTime, newUpperTime);
+            videoPos++;
+            Thread.sleep(100, 0);
```
Comment (aldebaran30701) [inline CoreTimelineView/src/au/gov/asd/tac/constellation/views/timeline/TimelineTopComponent.java:232]: Curious as to the need for the sleep calls?
Is this so the current frame is viewed for 100ms ?

File: CoreTimelineView/src/au/gov/asd/tac/constellation/views/timeline/TimelineTopComponent.java [hunk 2]
```diff
@@ -213,7 +208,70 @@ public void setExtents(final double lowerTimeExtent, final double upperTimeExten
...
+        originalLowerTime = getTimelineLowerTimeExtent();
+        originalUpperTime = getTimelineUpperTimeExtent();
+        videoParamsSet = true;
+
+        while (videoPos > 0 && videoParamsSet) {
+            double newLowerTime = (originalUpperTime - originalLowerTime) / 101 * videoPos + originalLowerTime;
+            double newUpperTime = (originalUpperTime - originalLowerTime) / 101 * (videoPos + 1) + originalLowerTime;
+            setExtents(newLowerTime, newUpperTime);
+            videoPos--;
+            Thread.sleep(100, 0);
```
Comment (aldebaran30701) [inline CoreTimelineView/src/au/gov/asd/tac/constellation/views/timeline/TimelineTopComponent.java:256]: sleep here also

**Review by aldebaran30701 — COMMENTED** [review]
Just a static look at the code, I had issues with my Netbeans and couldn't check this out and run it. (probably nothing to do with your branch).
I'm curious as to the sleeps, and if there is a better way of implementing whatever it actually does.

**Comment by arcturus2** [comment]
Just letting you know your other PR (https://github.com/constellation-app/constellation/pull/1453) has been merged in. Did you want to continue this one as well or close?

**Comment by CrucisGamma** [comment]
> Just letting you know your other PR (#1453) has been merged in. Did you want to continue this one as well or close?
Hey @arcturus2  Will close this one!

**Pull request closed without merging**